File: src/mrkto/resources/activity.py
"""Activity lookups and lead change tracking."""

from __future__ import annotations

from datetime import datetime, timedelta, timezone


def _since_datetime(days: int) -> str:
    since = datetime.now(timezone.utc) - timedelta(days=days)
    return since.strftime("%Y-%m-%dT%H:%M:%SZ")


def _paging_token(client, days: int) -> str:
    paging = client.get("/v1/activities/pagingtoken.json", params={"sinceDatetime": _since_datetime(days)})
    token = paging.get("nextPageToken")
    if not token:
        raise KeyError("Marketo paging token response did not include nextPageToken")
    return token


def get_activity_types(client) -> dict:
    return client.get("/v1/activities/types.json")


def list_activities(
    client,
    *,
    lead_id: int,
    activity_type_ids: list[int] | None = None,
    since_days: int = 30,
    limit: int | None = None,
) -> dict:
    params = {
        "leadIds": str(lead_id),
        "nextPageToken": _paging_token(client, since_days),
    }
    if activity_type_ids:
        params["activityTypeIds"] = ",".join(str(activity_type_id) for activity_type_id in activity_type_ids)
    return client.get_all_pages("/v1/activities.json", params=params, limit=limit)


def get_lead_changes(
    client,
    *,
    fields: list[str],
    since_days: int = 30,
    lead_ids: list[int] | None = None,
    list_id: int | None = None,
    limit: int | None = None,
) -> dict:
    params = {
        "fields": ",".join(fields),
        "nextPageToken": _paging_token(client, since_days),
    }
    if lead_ids:
        params["leadIds"] = ",".join(str(lead_id) for lead_id in lead_ids)
    if list_id is not None:
        params["listId"] = list_id
    return client.get_all_pages("/v1/activities/leadchanges.json", params=params, limit=limit)

File: internal/cmd/activity.go
package cmd

import (
	"fmt"
	"strconv"
	"strings"
	"time"

	"github.com/spf13/cobra"
)

func newActivityCmd(runtime *Runtime, options *RootOptions) *cobra.Command {
	activityCmd := &cobra.Command{
		Use:   "activity",
		Short: "Activity lookups and lead changes.",
		RunE: func(cmd *cobra.Command, args []string) error {
			return cmd.Help()
		},
	}

	activityCmd.AddCommand(newActivityTypesCmd(runtime, options))
	activityCmd.AddCommand(newActivityListCmd(runtime, options))
	activityCmd.AddCommand(newActivityChangesCmd(runtime, options))

	return activityCmd
}

func newActivityTypesCmd(runtime *Runtime, options *RootOptions) *cobra.Command {
	return &cobra.Command{
		Use:   "types",
		Short: "List Marketo activity types.",
		RunE: func(cmd *cobra.Command, args []string) error {
			apiClient, err := loadClient(runtime, options.Profile)
			if err != nil {
				return err
			}

			result, err := apiClient.Get("/v1/activities/types.json", nil)
			if err != nil {
				return err
			}

			return writeResult(runtime, options, result)
		},
	}
}

func newActivityListCmd(runtime *Runtime, options *RootOptions) *cobra.Command {
	var (
		typeIDs []int
		since   int
		limit   int
	)

	cmd := &cobra.Command{
		Use:   "list <lead-id>",
		Short: "List activities for a lead.",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			apiClient, err := loadClient(runtime, options.Profile)
			if err != nil {
				return err
			}

			pagingToken, err := activityPagingToken(apiClient, since)
			if err != nil {
				return err
			}

			params := map[string]any{
				"leadIds":       args[0],
				"nextPageToken": pagingToken,
			}
			if len(typeIDs) > 0 {
				params["activityTypeIds"] = joinInts(typeIDs)
			}

			result, err := apiClient.GetAllPages("/v1/activities.json", params, limit, 0)
			if err != nil {
				return err
			}

			return writeResult(runtime, options, result)
		},
	}

	flags := cmd.Flags()
	flags.IntSliceVar(&typeIDs, "type-id", nil, "Filter to activity type ids.")
	flags.IntVar(&since, "since", 30, "Days back from now.")
	flags.IntVar(&limit, "limit", 0, "Maximum number of records to return.")

	return cmd
}

func activityPagingToken(apiClient apiGetter, sinceDays int) (string, error) {
	result, err := apiClient.Get("/v1/activities/pagingtoken.json", map[string]any{
		"sinceDatetime": sinceDatetime(sinceDays),
	})
	if err != nil {
		return "", err
	}

	token, _ := result["nextPageToken"].(string)
	if token == "" {
		return "", fmt.Errorf("Marketo paging token response did not include nextPageToken")
	}
	return token, nil
}

func sinceDatetime(days int) string {
	return time.Now().UTC().Add(-time.Duration(days) * 24 * time.Hour).Format("2006-01-02T15:04:05Z")
}

func joinInts(values []int) string {
	parts := make([]string, 0, len(values))
	for _, value := range values {
		parts = append(parts, strconv.Itoa(value))
	}
	return strings.Join(parts, ",")
}

type apiGetter interface {
	Get(path string, params map[string]any) (map[string]any, error)
	GetAllPages(path string, params map[string]any, limit int, batchSize int) (map[string]any, error)
}

func newActivityChangesCmd(runtime *Runtime, options *RootOptions) *cobra.Command {
	var (
		watch   []string
		leadIDs []int
		listID  int
		since   int
		limit   int
	)

	cmd := &cobra.Command{
		Use:   "changes",
		Short: "List lead change activities.",
		RunE: func(cmd *cobra.Command, args []string) error {
			apiClient, err := loadClient(runtime, options.Profile)
			if err != nil {
				return err
			}

			pagingToken, err := activityPagingToken(apiClient, since)
			if err != nil {
				return err
			}

			params := map[string]any{
				"fields":        strings.Join(watch, ","),
				"nextPageToken": pagingToken,
			}
			if len(leadIDs) > 0 {
				params["leadIds"] = joinInts(leadIDs)
			}
			if cmd.Flags().Changed("list-id") {
				params["listId"] = listID
			}

			result, err := apiClient.GetAllPages("/v1/activities/leadchanges.json", params, limit, 0)
			if err != nil {
				return err
			}

			return writeResult(runtime, options, result)
		},
	}

	flags := cmd.Flags()
	flags.StringArrayVar(&watch, "watch", nil, "Lead field name to watch. Repeat the option for multiple fields.")
	flags.IntSliceVar(&leadIDs, "lead-id", nil, "Filter to specific lead ids.")
	flags.IntVar(&listID, "list-id", 0, "Filter to a static list id.")
	flags.IntVar(&since, "since", 30, "Days back from now.")
	flags.IntVar(&limit, "limit", 0, "Maximum number of records to return.")
	_ = cmd.MarkFlagRequired("watch")

	return cmd
}
